crates/holochain/src/core/workflow/integrate_dht_ops_workflow.rs

```diff
@@ -176,107 +160,130 @@ pub async fn integrate_dht_ops_workflow(
+        validation_status: iv.validation_status,
+        op: iv.op,
+        when_integrated: Timestamp::now(),
+    };
+    debug!("integrating");
+    Ok(Outcome::Integrated(integrated))
+}
+
 /// Check if we have the required dependencies held before integrating.
-// TODO: This doesn't really check why we are holding the values.
-// We could have them for other reasons.
-fn op_dependencies_held<P: PrefixType>(
+async fn op_dependencies_held(
     op: &DhtOp,
-    element_store: &ElementBuf<P>,
-) -> DhtOpConvertResult<bool> {
+    workspace: &mut IntegrateDhtOpsWorkspace,
+) -> CascadeResult<bool> {
     {
-        fn header_with_entry_is_stored<P: PrefixType>(
-            hash: &HeaderHash,
-            element_store: &ElementBuf<P>,
-        ) -> DhtOpConvertResult<bool> {
-            match element_store.get_header(hash)?.map(|e| {
-                e.header()
-                    .entry_data()
-                    .map(|(h, _)| h.clone())
-                    .ok_or_else(|| {
-                        // This is not a NewEntryHeader: cannot continue
-                        DhtOpConvertError::MissingEntryDataForHeader(hash.clone())
-                    })
-            }) {
-                Some(r) => Ok(element_store.contains_entry(&r?)?),
-                None => Ok(false),
-            }
-        }
-
-        // let entry_is_stored = |hash| element_store.contains_entry(hash);
-
-        let header_is_stored = |hash| element_store.contains_header(hash);
-
         match op {
-            DhtOp::StoreElement(_, _, _)
-            | DhtOp::StoreEntry(_, _, _)
-            | DhtOp::RegisterAgentActivity(_, _) => (),
+            DhtOp::StoreElement(_, _, _) | DhtOp::StoreEntry(_, _, _) => (),
+            DhtOp::RegisterAgentActivity(_, header) => {
+                // RegisterAgentActivity is the exception where we need to make
+                // sure that we have integrated the previous RegisterAgentActivity DhtOp
+                // from the same chain.
+                // This is because to say if the chain is valid as a whole we can't
+                // have parts missing.
+                let mut cascade = workspace.cascade();
+                let prev_header_hash = header.prev_header();
+                if let Some(prev_header_hash) = prev_header_hash.cloned() {
+                    match cascade
+                        .retrieve_header(prev_header_hash, Default::default())
+                        .await?
+                    {
+                        Some(prev_header) => {
+                            let op_hash = DhtOpHash::with_data_sync(
+                                &UniqueForm::RegisterAgentActivity(prev_header.header()),
+                            );
+                            if workspace.integration_limbo.contains(&op_hash)? {
+                                return Ok(true);
+                            }
+                        }
+                        None => return Ok(false),
```

Comment:
Just a style thing: early returns make me wonder what other branch of logic I might be missing, and this is a case where they could probably be removed to let the whole function return all at once. But it's such a short function (nice!) that it doesn't make that much difference in this case.